### PR TITLE
Feature: add import init cluster support

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Feature: add import init cluster support

## Changelog

- New Feature
- Documentation

## Related Issues

Fixes [Issue #6279](https://github.com/datafuselabs/databend/issues/6279)

### CODE REVIEW

File: docs/doc/50-manage/00-metasrv/30-metasrv-backup-restore.md [hunk 1]
```diff
@@ -27,13 +27,13 @@ E.g., every line in the output file is a JSON of an exported key-value record.
...
 
-##  Restore a databend-meta
+## Restore a databend-meta
 
 The following command rebuild a meta service db in `<your_meta_dir>` from
 exported metadata:
 
 ```sh
-cat "<output_fn>" | ./target/debug/databend-metactl --import --raft-dir "<your_meta_dir>"
+./target/debug/databend-metactl --import --raft-dir "<your_meta_dir>" --db <output_fn>
```
Comment (ZhiHanZ) [inline docs/doc/50-manage/00-metasrv/30-metasrv-backup-restore.md:36]: could you elaborate on how could we restore a three node new cluster from one snapshot, ie
```bash
./target/debug/databend-metactl --import --raft-dir "<your_meta_dir1>" --db <output_fn>  --initial-cluster 1=localhost:28003
./target/debug/databend-metactl --import --raft-dir "<your_meta_dir2>" --db <output_fn>  --initial-cluster 2=localhost:28004
./target/debug/databend-metactl --import --raft-dir "<your_meta_dir3>" --db <output_fn>  --initial-cluster 3=localhost:28005

// start cluster
./target/debug/databend-meta --raft-dir "your raft dir1" --rpc-advertise-host localhost:28003 --id=1
./target/debug/databend-meta --raft-dir "your raft dir2" --rpc-advertise-host localhost:28004 --id=2
./target/debug/databend-meta --raft-dir "your raft dir3" --rpc-advertise-host localhost:28005 --id=3

```

File: tests/metactl/test-metactl.sh [hunk 2]
```diff
@@ -48,3 +50,16 @@ echo " === check if there is a node record in it"
 grep -Fxq '["raft_state",{"RaftStateKV":{"key":"Id","value":{"NodeId":0}}}]' $grpc_exported
 
 kill $METASRV_PID
+
+echo " === export data from a running databend-meta to $grpc_exported with initial-cluster args"
+./target/${BUILD_PROFILE}/databend-metactl --import --raft-dir $import_meta_dir --db $grpc_exported --initial-cluster 3=http://localhost:23822,0.0.0.0:9292
```
Comment (ZhiHanZ) [inline tests/metactl/test-metactl.sh:55]: I think we should test on whether we could retrieve the same result from new cluster recovered from metactl

**Comment by ZhiHanZ** [comment]
do we need to configure initial-raft-advertise-host to establish identity?

File: common/meta/raft-store/src/state_machine/sm.rs [hunk 3]
```diff
@@ -298,13 +298,13 @@ impl StateMachine {
...
                 EntryPayload::Membership(ref mem) => {
-                    txn_sm_meta.insert(
+                    let r = txn_sm_meta.insert(
                         &LastMembership,
                         &StateMachineMetaValue::Membership(EffectiveMembership {
                             log_id: *log_id,
                             membership: mem.clone(),
                         }),
-                    )?;
+                    );
```
Comment (drmingdrmer) [inline common/meta/raft-store/src/state_machine/sm.rs:307]: What's the purpose to remove the result checking with `?`?
`insert()` may emit a StorageError that should be popped up to notify the the server to shutdown at once.

File: tools/metactl/src/snapshot.rs [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    return Ok(());
+}
+
+pub async fn import_data(config: &Config) -> anyhow::Result<()> {
+    let raft_config = &config.raft_config;
+    eprintln!("import meta dir into: {}", raft_config.raft_dir);
+
+    init_sled_db(raft_config.raft_dir.clone());
+
+    clear()?;
```
Comment (drmingdrmer) [inline tools/metactl/src/snapshot.rs:73]: 👍

File: tools/metactl/src/snapshot.rs [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        tree.insert(k, v)?;
+        n += 1;
+
+        match kv_variant {
+            KeySpaceKV::Logs { key: _, value } => {
+                if value.log_id > max_log_id {
+                    max_log_id = value.log_id;
+                }
+            }
+            _ => {}
```
Comment (drmingdrmer) [inline tools/metactl/src/snapshot.rs:112]: There is a chance there is no log at all, e.g., when a snapshot is installed that overrides all logs.

In this case it should find out the max log id from state machine.

File: common/meta/sled-store/src/sled_tree.rs [hunk 6]
```diff
@@ -641,6 +642,14 @@ impl<'a, KV: SledKeySpace> AsKeySpace<'a, KV> {
         self.inner.range_remove::<KV, R>(range, flush).await
     }
 
+    pub fn clear(&self) -> MetaStorageResult<()> {
+        let err = self.inner.tree.clear();
+        match err {
+            Err(err) => Err(MetaStorageError::SledError(AnyError::new(&err))),
```
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply common/meta/sled-store/src/sled_tree.rs:648]: anyhow can not be serialized and it does not impl `Error`.

File: tests/metactl/test-metactl.sh (hunk 2 above)
Comment (lichuang) [reply tests/metactl/test-metactl.sh]: add test-metactl-restore-new-cluster.sh to test this case.

File: common/meta/raft-store/src/state/raft_state.rs [hunk 7]
```diff
@@ -104,17 +104,22 @@ impl RaftState {
...
     #[tracing::instrument(level = "debug", skip(self))]
-    async fn init(&self) -> MetaResult<()> {
+    pub async fn set_cluster_id(&self, id: NodeId) -> MetaResult<()> {
         let state = self.state();
         state
-            .insert(&RaftStateKey::Id, &RaftStateValue::NodeId(self.id))
+            .insert(&RaftStateKey::Id, &RaftStateValue::NodeId(id))
             .await?;
         Ok(())
     }
```
Comment (drmingdrmer) [inline common/meta/raft-store/src/state/raft_state.rs:114]: Why `cluster_id`?

It is storing a `node-id` IMHO.

File: tools/metactl/src/snapshot.rs [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+            log.insert(&entry).await?;
+        }
+    }
+
+    raft_state.set_cluster_id(id).await?;
+    match last_applied_log_id {
+        Some(id) => sm.set_last_applied(id).await?,
+        None => {}
+    }
+    Ok(())
```
Comment (drmingdrmer) [inline tools/metactl/src/snapshot.rs:254]: `last_applied_log_id` is read from the state machine and is not changed at all. I do not think it is necessary to update it.

File: common/meta/raft-store/src/state_machine/sm.rs [hunk 9]
```diff
@@ -847,6 +847,20 @@ impl StateMachine {
         Ok(last_applied)
     }
 
+    pub async fn set_last_applied(&self, last_applied: LogId) -> MetaStorageResult<()> {
+        let sm_meta = self.sm_meta();
+        sm_meta
+            .insert(&LastApplied, &StateMachineMetaValue::LogId(last_applied))
+            .await?;
+        Ok(())
+    }
```
Comment (drmingdrmer) [inline common/meta/raft-store/src/state_machine/sm.rs:856]: It can be removed now right?

File: tests/metactl/test-metactl.sh [hunk 10]
```diff
@@ -3,48 +3,64 @@
...
-echo " === exported file data end"
+# test export from grpc
+chmod +x ./target/debug/databend-meta
+./target/debug/databend-meta --single &
+METASRV_PID=$!
+echo $METASRV_PID
+sleep 0.5
 
-echo " === check if there is a node record in it"
+./target/debug/databend-metactl --export --grpc-api-address "127.0.0.1:9191"  > $grpc_exported
```
Comment (drmingdrmer) [inline tests/metactl/test-metactl.sh:63]: Why removed the `echo`s...

They will help a lot when your CI fails.

File: tools/metactl/src/snapshot.rs [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+
+    if config.initial_cluster.is_empty() || config.raft_config.id == 0 {
+        return Ok(());
+    }
+    return init_new_cluster(
+        config.initial_cluster.clone(),
+        max_log_id,
+        config.raft_config.id,
+    )
+    .await;
```
Comment (drmingdrmer) [inline tools/metactl/src/snapshot.rs:84]: The returned error should be popped up.

**Pull request merged**